modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotCheckProcess.java

```diff
@@ -0,0 +1,501 @@
+                if (err != null)
+                    phaseFut.onDone(err);
+                else
+                    phaseFut.onDone(new SnapshotCheckResponse(res));
+            });
+        }
+
+        return phaseFut;
+    }
+
+    /** */
+    private Map<ClusterNode, Exception> mapErrors(Map<UUID, Throwable> errors) {
+        return errors.entrySet().stream()
+            .collect(Collectors.toMap(e -> kctx.cluster().get().node(e.getKey()),
+                e -> e.getValue() instanceof Exception ? (Exception)e.getValue() : new IgniteException(e.getValue())));
+    }
+
+    /** */
+    private Map<ClusterNode, Map<PartitionKeyV2, PartitionHashRecordV2>> mapPartsHashes(
+        Map<UUID, SnapshotCheckResponse> results,
+        Collection<UUID> requiredNodes
+    ) {
+        return results.entrySet().stream()
+            .filter(e -> requiredNodes.contains(e.getKey()) && e.getValue() != null)
+            .collect(Collectors.toMap(e -> kctx.cluster().get().node(e.getKey()), e -> e.getValue().partsHashes()));
+    }
+
+    /** */
+    private Map<ClusterNode, Map<String, SnapshotHandlerResult<?>>> mapCustomHandlersResults(
+        Map<UUID, SnapshotCheckResponse> results,
+        Set<UUID> requiredNodes
+    ) {
+        return results.entrySet().stream()
+            .filter(e -> requiredNodes.contains(e.getKey()) && e.getValue() != null)
+            .collect(Collectors.toMap(e -> kctx.cluster().get().node(e.getKey()), e -> e.getValue().customHandlersResults()));
+    }
+
+    /**
+     * @param snpName Snapshot name of the validation process. If {@code null}, ignored.
+     * @param reqId  If {@code snpName} is {@code null}, is used to find the operation request.
+     * @return Current snapshot checking context by {@code snpName} or {@code reqId}.
+     */
+    private @Nullable SnapshotCheckContext context(@Nullable String snpName, UUID reqId) {
+        SnapshotCheckContext ctx = snpName == null
+            ? contexts.values().stream().filter(ctx0 -> ctx0.req.requestId().equals(reqId)).findFirst().orElse(null)
+            : contexts.get(snpName);
+
+        assert ctx == null || ctx.req.requestId().equals(reqId);
+
+        return ctx;
+    }
+
+    /** Phase 1 beginning: prepare, collect and check local metas. */
+    private IgniteInternalFuture<SnapshotCheckResponse> prepareAndCheckMetas(SnapshotCheckProcessRequest req) {
+        if (!req.nodes().contains(kctx.localNodeId()))
+            return new GridFinishedFuture<>();
+
+        if (kctx.isStopping())
+            return new GridFinishedFuture<>(new NodeStoppingException("The node is stopping: " + kctx.localNodeId()));
+
+        SnapshotCheckContext ctx = contexts.computeIfAbsent(req.snapshotName(), snpName -> new SnapshotCheckContext(req));
+
+        if (!ctx.req.requestId().equals(req.requestId())) {
+            return new GridFinishedFuture<>(new IllegalStateException("Validation of snapshot '" + req.snapshotName()
```

Comment:
Just throw exception here

Reply:
We've already discussed that. We do not throws exceptions on another DPs. We return futures. Additional error messages are posted in the log in case of direct exception in Distributed process:

` catch (Throwable err) {
                U.error(log, "Failed to handle InitMessage [id=" + p.id + ']', err);
`